arngll/src/frame_info.rs

```diff
@@ -0,0 +1,690 @@
+            iter.next().copied().unwrap(),
+            iter.next().copied().unwrap(),
+            iter.next().copied().unwrap(),
+            iter.next().copied().unwrap()
+        ]);
+        let kid = if kim == KeyIdentMode::KeyIndex {
+            Some(iter.next().copied().unwrap())
+        } else {
+            None
+        };
+
+        SecInfo {
+            enc,
+            kim,
+            fcntr,
+            kid,
+            mic: Mic { len: miclen, .. Mic::EMPTY }
+        }
+    }
+
+    pub fn scf(&self) -> u8 {
+        (u8::from(self.enc) << 7)
+            + (self.mic.len.to_u8() << 5)
+            + (self.kim.to_u8() << 3)
+    }
+
+    pub fn bytes(&self) -> impl Iterator<Item=u8> {
+        once(self.scf())
+            .chain(self.fcntr.to_be_bytes())
+            .chain(self.kid)
+    }
+}
+
+impl Debug for SecInfo {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{{")?;
+        if self.enc {
+            write!(f, "ENC ")?;
+        }
+
+        write!(f, "FCNTR={} KIM={:?}",self.fcntr, self.kim)?;
+
+        if let Some(kid) = self.kid {
+            write!(f, " KID=0x{:02X}",kid)?;
+        }
+
+        write!(f, " MIC={:?}",self.mic)?;
+        write!(f, "}}")
+    }
+}
+
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct FrameInfo {
+    pub frame_type: FrameType,
+    pub ack_requested: bool,
+    pub is_from_relay: bool,
+    pub network_id: Option<NetworkId>,
+    pub dst_addr: HamAddr,
+    pub src_addr: HamAddr,
+    pub rly_addr: Option<HamAddr>,
+    pub sec_info: Option<SecInfo>,
+    pub ack_crc: u16,
+}
+
+impl Debug for FrameInfo {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f,"{{{:?}", self.frame_type)?;
+
+        if self.ack_requested {
+            write!(f," AckReq")?;
+        }
+
+        if self.is_from_relay {
+            write!(f," FromRly")?;
+        }
+
+        if let Some(netid) = self.network_id {
+            write!(f," NetId={:?}", netid)?;
+        }
+
+        if !self.dst_addr.is_empty() {
+            write!(f," Dst={}", self.dst_addr)?;
+        }
+
+        write!(f," Src={}", self.src_addr)?;
+
+        if self.frame_type == FrameType::Ack {
+            write!(f," AckCrc=0x{:04X}", self.ack_crc)?;
+        }
+
+        if let Some(rly_addr) = self.rly_addr {
+            write!(f," Rly={}", rly_addr)?;
+        }
+
+        if let Some(secinfo) = &self.sec_info {
+            write!(f," SecInfo={:?}", secinfo)?;
+        }
+
+        write!(f,"}}")
+    }
+}
+
+impl FrameInfo {
+    pub const EMPTY: FrameInfo = FrameInfo {
+        frame_type: FrameType::Data,
+        ack_requested: false,
+        is_from_relay: false,
+        network_id: None,
+        dst_addr: HamAddr::EMPTY,
+        src_addr: HamAddr::EMPTY,
+        rly_addr: None,
+        sec_info: None,
+        ack_crc: 0,
+    };
+
+    pub fn ack_calc(&self, payload: &[u8]) -> Option<(u16,HamAddr)> {
+        if self.ack_requested {
```

Comment:
I believe this could be expressed as, `self.ack_requested.then_some({ ... })`, but that may be a bit too clever.